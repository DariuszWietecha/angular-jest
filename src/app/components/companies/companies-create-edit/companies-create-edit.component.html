<h4><span class="text-capitalize">{{ type }}</span> company</h4>

<form [formGroup]="createEditForm" (ngSubmit)="onSubmit()">
  <div>
    <dt>
      <label for="name">
        Name
      </label>
    </dt>
    <dd>
      <input id="name" type="text" formControlName="name">
    </dd>
  </div>
  <div>
    <dt>
      <label for="logo-url">
        LogoUrl
      </label>
    </dt>
    <dd>
      <input id="logo-url" type="text" formControlName="logoUrl">
    </dd>
  </div>
  <div>
    <dt>
      <label for="e-mail">
        E-mail
      </label>
    </dt>
    <dd>
      <input id="e-mail" type="text" formControlName="email">
    </dd>
  </div>
  <div>
    <dt>
      <label for="categories">
        Categories
      </label>
    </dt>
    <dd>
      <select id="categories" multiple formControlName="categories">
      <option *ngFor="let category of categories"
      [ngValue]="category"
      [selected]="isCategorySelected(category.id)">
        {{category.name}}</option>
    </select>
    <div><small>Use Ctrl + left mouse button to select multiple items or unselect all.</small></div>
  </dd>
  </div>

  <button class="button" type="submit" class="btn btn-primary">Save</button>
</form>