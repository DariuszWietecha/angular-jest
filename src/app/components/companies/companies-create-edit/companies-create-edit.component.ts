import { Component, OnInit, OnDestroy } from '@angular/core';
import { ActivatedRoute } from '@angular/router';
import { FormBuilder, FormGroup } from '@angular/forms';
import { Router } from '@angular/router';
import { Subscription } from 'rxjs';
import { Store, ActionsSubject } from '@ngrx/store';
import { IStore, selectEditedCompany, selectCategoriesData } from '../../../reducers';
import * as companiesActions from '../../../actions/companies.actions';
import * as categoriesActions from '../../../actions/categories.actions';
import { take } from 'rxjs/operators';
import { ofType } from '@ngrx/effects';

@Component({
  selector: 'app-companies-create-edit',
  templateUrl: './companies-create-edit.component.html',
  styleUrls: ['./companies-create-edit.component.scss']
})
export class CompaniesCreateEditComponent implements OnInit, OnDestroy {
  actionSub = new Subscription();
  categories = [];
  createEditForm: FormGroup;
  routeSub = new Subscription();
  type: string;

  isCategorySelected = (categoryId) => this.createEditForm.value.categories &&
    this.createEditForm.value.categories.some((category) => category.id === categoryId)

  constructor(
    private actionsSubject: ActionsSubject,
    private route: ActivatedRoute,
    private formBuilder: FormBuilder,
    private router: Router,
    private store: Store<IStore>
  ) {
    this.createEditForm = this.formBuilder.group({
      id: '',
      name: '',
      logoUrl: '',
      email: '',
      categories: [],
    });

    this.actionSub = actionsSubject.pipe(
      ofType(companiesActions.createCompanySuccess, companiesActions.editCompanySuccess)
    ).subscribe(() => {
      this.router.navigate(['/companies']);
    });
  }

  ngOnInit() {
    this.routeSub = this.route.paramMap.subscribe(params => {
      this.store.select(selectCategoriesData)
        .pipe(take(2))
        .subscribe((data) => {
          this.categories = data;
        });
      this.store.dispatch(categoriesActions.loadCategories());

      this.type = params.get('type');
      if (this.type === 'edit') {
        this.store.select(selectEditedCompany)
          .pipe(take(2))
          .subscribe((data) => this.createEditForm.patchValue(data));
        this.store.dispatch(companiesActions.loadCompany({ companyId: params.get('id') }));
      }
    });
  }

  onSubmit() {
    const action = this.type === 'edit' ? 'editCompany' : 'createCompany';
    const formData = this.createEditForm.value;
    if (!formData.id) {
      delete formData.id;
    }

    this.store.select(selectEditedCompany)
      .pipe(take(1))
      .subscribe((data) => this.createEditForm.patchValue(data));

    this.store.dispatch(companiesActions[action]({ data: formData }));

    this.createEditForm.reset();
  }

  ngOnDestroy() {
    this.actionSub.unsubscribe();
    this.routeSub.unsubscribe();
  }
}
