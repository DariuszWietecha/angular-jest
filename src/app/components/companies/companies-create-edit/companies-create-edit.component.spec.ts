import { CompaniesCreateEditComponent } from './companies-create-edit.component';
import { ReactiveFormsModule } from '@angular/forms';
import { CategoriesService } from '../../../services/categories.service';
import { CompaniesService } from '../../../services/companies.service';
import { HttpClientModule } from '@angular/common/http';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { RouterTestingModule } from '@angular/router/testing';

describe('CompaniesCreateEditComponent', () => {
  let component: CompaniesCreateEditComponent;
  let fixture: ComponentFixture<CompaniesCreateEditComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [ HttpClientModule, ReactiveFormsModule, RouterTestingModule],
      declarations: [ CompaniesCreateEditComponent ],
      providers: [CategoriesService, CompaniesService],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CompaniesCreateEditComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('snapShot', () => {
    expect(fixture).toMatchSnapshot();
  });
});
