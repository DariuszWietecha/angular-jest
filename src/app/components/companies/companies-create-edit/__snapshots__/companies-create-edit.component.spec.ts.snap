// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompaniesCreateEditComponent snapShot 1`] = `
<app-companies-create-edit
  categories={[Function Array]}
  categoriesService={[Function CategoriesService]}
  companiesService={[Function CompaniesService]}
  createEditForm={[Function FormGroup]}
  formBuilder={[Function FormBuilder]}
  isCategorySelected={[Function Function]}
  route={[Function ActivatedRoute]}
  router={[Function Router]}
  type="null"
>
  <h4>
    <span
      class="text-capitalize"
    >
      
    </span>
     company
  </h4><form
    class="ng-untouched ng-pristine ng-valid"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    <div>
      <dt>
        <label
          for="name"
        >
           Name 
        </label>
      </dt>
      <dd>
        <input
          class="ng-untouched ng-pristine ng-valid"
          formcontrolname="name"
          id="name"
          ng-reflect-name="name"
          type="text"
        />
      </dd>
    </div>
    <div>
      <dt>
        <label
          for="logo-url"
        >
           LogoUrl 
        </label>
      </dt>
      <dd>
        <input
          class="ng-untouched ng-pristine ng-valid"
          formcontrolname="logoUrl"
          id="logo-url"
          ng-reflect-name="logoUrl"
          type="text"
        />
      </dd>
    </div>
    <div>
      <dt>
        <label
          for="e-mail"
        >
           E-mail 
        </label>
      </dt>
      <dd>
        <input
          class="ng-untouched ng-pristine ng-valid"
          formcontrolname="email"
          id="e-mail"
          ng-reflect-name="email"
          type="text"
        />
      </dd>
    </div>
    <div>
      <dt>
        <label
          for="categories"
        >
           Categories 
        </label>
      </dt>
      <dd>
        <select
          class="ng-untouched ng-pristine ng-valid"
          formcontrolname="categories"
          id="categories"
          multiple=""
          ng-reflect-name="categories"
        >
          
        </select>
        <div>
          <small>
            Use Ctrl + left mouse button to select multiple items or unselect all.
          </small>
        </div>
      </dd>
    </div>
    <button
      class="btn btn-primary"
      type="submit"
    >
      Save
    </button>
  </form>
</app-companies-create-edit>
`;
